_posts/2022-09-06-Umesh-GSoC-on-ConscryptMode.md

```diff
@@ -0,0 +1,24 @@
+---
+title:  "GSoC 2022 final submission for ConscryptMode"
+author: Umesh-01
+---
+
+My name is Umesh Singh and I was an open source contributor through Google Summer of Code this year. Google Summer of Code (GSoC) is a program where external participants can contribute to an open source project over a few months. We learn new computer science concepts, how to work on open source repositories, and create real code contributions to projects!
+
+My project was `Switching Robolectric from BouncyCastle to Conscrypt as the default security provider`. Robolectric was using [BouncyCastle](https://www.bouncycastle.org/) as the Java Cryptography Extension (JCE) security provider. After the introduction of Android P, Android switched to using [Google Conscrypt](https://source.android.com/docs/core/architecture/modular-system/conscrypt) as the security provider. 
+To be more consistent with Android, Robolectric needed to be updated to use Conscrypt as the default security provider.
+ 
+I was already well familiar with the technologies used in the project such as Java, Kotlin, Android Studio, Git and GitHub. I was interested in Android Developement and I was looking for an android project that uses Java as the primary language. I found Robolectric to be a perfect project to contribute to based on my skills, and I was ready to learn about testing concepts since Robolectric is a popular unit testing framework for Android. 
+
+At first, we thought we could switch Robolectric from BouncyCastle to Conscrypt directly. We quickly found we can't completely switch Robolectric to Conscrypt because there were some internal tests that needed BouncyCastle as a fallback security provider. After we added it in, we decided to create a ConscryptMode annotation that lets users choose to use Conscrypt or not. Robolectric can now use either Conscrypt and BouncyCastle or just BouncyCastle as the security provider. Robolectric will search for a requested security feature from Conscrypt first. If it does not support it, Robolectric will try BouncyCastle second.
```

Comment:
Change this to:

The initial approach for this project was to drop BouncyCastle entirely and switch to Conscrypt. This seemed appealing due to the prior issues with BouncyCastle, such as [#5456](https://github.com/robolectric/robolectric/issues/5456). However, we discovered that we couldn't completely switch Robolectric to Conscrypt. Conscrypt only supports the security primitives provided by BoringSSL, and there were some legacy security primitives that were still being used in Android. Because of this we realized that BouncyCastle was needed as the fallback security provider. We settled on having a `@ConscryptMode` annotation that lets users choose whether Conscrypt will be installed or not. When Conscrypt is enabled, Robolectric will search for a requested security feature from Conscrypt first. If it does not support it, the other security providers will be queried. This is more consistent with how Android does it.